velox/functions/lib/aggregates/AverageAggregateBase.h

```diff
@@ -0,0 +1,417 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "velox/exec/Aggregate.h"
+#include "velox/expression/FunctionSignature.h"
+#include "velox/functions/prestosql/aggregates/AggregateNames.h"
+#include "velox/functions/prestosql/aggregates/DecimalAggregate.h"
+#include "velox/type/DecimalUtil.h"
+#include "velox/vector/ComplexVector.h"
+#include "velox/vector/DecodedVector.h"
+#include "velox/vector/FlatVector.h"
+
+namespace facebook::velox::functions::aggregate {
+
+namespace {
+
+/// Translate selected rows of decoded to the corresponding rows of its base
+/// vector.
+SelectivityVector translateToInnerRows(
+    const DecodedVector& decoded,
+    const SelectivityVector& rows) {
+  VELOX_DCHECK(!decoded.isIdentityMapping());
+  if (decoded.isConstantMapping()) {
+    auto constantIndex = decoded.index(rows.begin());
+    SelectivityVector baseRows{constantIndex + 1, false};
+    baseRows.setValid(constantIndex, true);
+    baseRows.updateBounds();
+    return baseRows;
+  } else {
+    SelectivityVector baseRows{decoded.base()->size(), false};
+    rows.applyToSelected(
+        [&](auto row) { baseRows.setValid(decoded.index(row), true); });
+    baseRows.updateBounds();
+    return baseRows;
+  }
+}
+
+/// Return the selected rows of the base vector of decoded corresponding to
+/// rows. If decoded is not identify mapping, baseRowsHolder contains the
+/// selected base rows. Otherwise, baseRowsHolder is unset.
+const SelectivityVector* getBaseRows(
+    const DecodedVector& decoded,
+    const SelectivityVector& rows,
+    SelectivityVector& baseRowsHolder) {
+  const SelectivityVector* baseRows = &rows;
+  if (!decoded.isIdentityMapping() && rows.hasSelections()) {
+    baseRowsHolder = translateToInnerRows(decoded, rows);
+    baseRows = &baseRowsHolder;
+  }
+  return baseRows;
+}
+
+template <typename TSum>
+struct SumCount {
+  TSum sum{0};
+  int64_t count{0};
+};
+
+} // namespace
+
+/// Partial aggregation produces a pair of sum and count.
+/// Count is BIGINT() while sum  and the final aggregates type depends on
+/// the input types:
+///       INPUT TYPE    |     SUM             |     AVG
+///   ------------------|---------------------|------------------
+///     REAL            |     DOUBLE          |    REAL
```

Comment:
This comment is no longer accurate, right?

Reply:
Moved this part to to be above `registerAverage(const std::string& name)` function of presto and spark sql.